### PR TITLE
Implement `KeplrError` and replace error instances in `@keplr-wallet/background`

### LINKED ISSUE #367: `KeplrError` type 추가하기
#314 

현재 오류를 처리할 때 따로 오류 타입을 만들지 않고 기본 Error만 사용해서 하고 있다.
하지만 #314 PR의 경우와 같이 가끔씩 특정 오류에 대해서는 다른 처리가 필요한 경우가 있다.
현재는 error.message를 직접 확인하는 방식으로 하고 있지만 이 방식은 관리하기 어렵다.

### TODO
- `KeplrError` 타입을 추가한다.
- `KeplrError` 타입은 code: number와 module: string을 필드로 가진다.
- background router와 message requester에서 `KeplrError` 오류의 경우 잘 처리해주도록 만든다.

### PR DESCRIPTION
**Pull request by junhoyeo**
- Closes #367 

<details>
  <summary>Complete table of error messages by each module</summary>

### Keyring
| module | code | message |
|:---:|:---:|:---|
| `keyring` | 100 | Invalid account in hd path |
| `keyring` | 101 | Invalid address index in hd path |
| `keyring` | 102 | Invalid change in hd path |
| `keyring` | 110 | Coin type already set |
| `keyring` | 111 | Invalid coin type passed in to Ethereum signing (expected 60) |
| `keyring` | 112 | Ethereum signing with Ledger is not yet supported |
| `keyring` | 120 | Invalid keystore |
| `keyring` | 121 | Invalid password |
| `keyring` | 130 | Key Store is empty (merged with `Empty key store`) |
| `keyring` | 131 | Key store's id is empty |
| `keyring` | 132 | Invalid type of key store |
| `keyring` | 133 | Key store type is mnemonic and it is unlocked. But, mnemonic is not loaded unexpectedly |
| `keyring` | 134 | Key store type is private key and it is unlocked. But, private key is not loaded unexpectedly |
| `keyring` | 140 | Key ring is locked (merged with `Keyring is locked`) |
| `keyring` | 141 | Key ring is locked or not initialized |
| `keyring` | 142 | Key ring is not loaded or not empty |
| `keyring` | 143 | Key ring is not unlocked |
| `keyring` | 144 | Key ring not initialized |
| `keyring` | 145 | Unexpected type of keyring |
| `keyring` | 150 | Ledger public key not set |
| `keyring` | 151 | Ledger public key is not initialized |
| `keyring` | 152 | Ledger is not compatible with this coinType right now |
| `keyring` | 200 | Index is negative |
| `keyring` | 201 | Invalid index |
| `keyring` | 202 | Invalid kdf |
| `keyring` | 210 | Pub key unmatched |
| `keyring` | 211 | `Unsupported type of pub key: ${signature.pub_key.type}` |
| `keyring` | 220 | Unknown kdf |
| `keyring` | 221 | Unknown msg type |
| `keyring` | 222 | Unmatched mac |
| `keyring` | 230 | signer not set |
| `keyring` | 231 | Signer mismatched |
| `keyring` | 232 | Unmatched signer in new sign doc |
| `keyring` | 233 | Unmatched signer in sign doc |
| `keyring` | 234 | Chain id in the message is not matched with the requested chain id |
| `keyring` | 235 | Sign options are null |
| `keyring` | 236 | Sign doc is not for ADR-36. But, "isADR36WithString" option is defined |
| `keyring` | 237 | Signing request was for ADR-36. But, accidentally, new sign doc is not for ADR-3 |
| `keyring` | 240 | coin type can not be negative |
| `keyring` | 241 | coin type should be integer |
| `keyring` | 250 | empty bip44 path list |
| `keyring` | 260 | invalid length of private key |
| `keyring` | 261 | key doesn't exist |
| `keyring` | 270 | chain id not set |
| `keyring` | 271 | Signature not set |
| `keyring` | 272 | mnemonic not set |
| `keyring` | 273 | name not set |
| `keyring` | 274 | password not set |
| `keyring` | 275 | private key not set |

### ledger
| module | code | message |
|:---:|:---:|:---|
| `ledger` | 100 | Cosmos App not initialized |
| `ledger` | 101 | Flag is null |
| `ledger` | 102 | Device is on screen saver |
| `ledger` | 110 | Unmatched public key |
| `ledger` | 111 | Unknown msg type |
| `ledger` | 112 | `Unknown mode: ${mode}` |
| `ledger` | 120 | Ledger init aborted |
| `ledger` | 121 | Ledger init timeout |

### chains
| module | code | message |
|:---:|:---:|:---|
| `chains` | 100 | Chain info not set |
| `chains` | 101 | Chain id not set |
| `chains` | 110 | Unknown msg type |
| `chains` | 111 | `There is no chain info for ${chainId}` |
| `chains` | 120 | Chain is not registered |
| `chains` | 121 | Same chain is already registered |
| `chains` | 122 | Can't remove the embedded chain |
| `chains` | 130 | cosmwasm and secretwasm are not compatible |

### interaction
| module | code | message |
|:---:|:---:|:---|
| `interaction` | 100 | Id is aleady in use |
| `interaction` | 101 | Type should not be empty |
| `interaction` | 110 | Unknown msg type |

### permission
| module | code | message |
|:---:|:---:|:---|
| `permission` | 100 | chain id not set |
| `permission` | 101 | origin not set |
| `permission` | 110 | empty permission type |
| `permission` | 111 | empty permission origin |
| `permission` | 120 | Unknown msg type |
| `permission` | 130 | `${origin} is not permitted` |

### tokens
| module | code | message |
|:---:|:---:|:---|
| `tokens` | 100 | Chain id is empty |
| `tokens` | 101 | Contract address is empty |
| `tokens` | 110 | Unknown type of currency |
| `tokens` | 111 | There is no matched secret20 |
| `tokens` | 120 | Unknown msg type |

### updater
| module | code | message |
|:---:|:---:|:---|
| `updater` | 100 | Empty chain id |
| `updater` | 110 | Unknown msg type |

### tx
| module | code | message |
|:---:|:---:|:---|
| `tx` | 100 | chain id is empty |
| `tx` | 101 | tx is empty |
| `tx` | 110 | Unknown msg type |
| `tx` | 120 | invalid mode |

### secret-wasm
| module | code | message |
|:---:|:---:|:---|
| `secret-wasm` | 100 | chain id not set |
| `secret-wasm` | 101 | msg not set |
| `secret-wasm` | 102 | ciphertext not set |
| `secret-wasm` | 103 | contract code hash not set |
| `secret-wasm` | 110 | nonce not set |
| `secret-wasm` | 111 | nonce is null |
| `secret-wasm` | 120 | Unknown msg type |
| `secret-wasm` | 130 | Key ring is not initialized |

</details>

### CODE REVIEW

File: packages/router/src/interfaces.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
 export interface Result {
-  error?: string;
+  error?: string | { module: string; code: number; message: string };
```
Comment (Thunnini) [inline packages/router/src/interfaces.ts:2]: { module: string; code: number; message: string }일 땐 KeplrError인 것으로 간주하고 처리되어야 한다고 주석에 적어주세요

File: packages/router/src/error.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export class KeplrError extends Error {
+  module: string;
+  code: number;
```
Comment (Thunnini) [inline packages/router/src/error.ts:3]: module이랑 code에 public readonly 추가해주세요

File: packages/stores/src/account/base.ts [hunk 3]
```diff
@@ -268,7 +269,7 @@ export class AccountSetBase {
     return this._walletStatus;
   }
 
-  get rejectionReason(): Error | undefined {
+  get rejectionReason(): Error | KeplrError | undefined {
```
Comment (Thunnini) [inline packages/stores/src/account/base.ts:272]: 제 생각엔 이미 Error 타입이 있기 때문에 KeplrError까지 union type으로 할 필요는 없을 것 같아요

File: packages/extension/src/pages/main/account.tsx [hunk 4]
```diff
@@ -60,9 +61,9 @@ export const AccountView: FunctionComponent = observer(() => {
         <ToolTip
           tooltip={(() => {
             if (
-              accountInfo.rejectionReason &&
-              accountInfo.rejectionReason.message ===
-                "Ledger is not compatible with this coinType right now"
+              accountInfo.rejectionReason instanceof KeplrError &&
+              accountInfo.rejectionReason.module === "keyring" &&
+              accountInfo.rejectionReason.code === 152
```
Comment (Thunnini) [inline packages/extension/src/pages/main/account.tsx:66]: if (**accountInfo.rejectionReason** && ...) 부분이 살아나야할 것 같아요.
undefined instanceof KeplrError는 false라서 없어도 될 것 같기는한데 null check이 없으니 찝찝해요

File: packages/stores/src/account/base.ts (hunk 3 above)
Comment (Thunnini) [reply packages/stores/src/account/base.ts:272]: 흠 이 부분이 애매한게 KeplrError이 union type으로 되어있어도 코드 상에서 역할은 없는 것 같긴해요... 근데 그냥 라이브러리 사용하는 사람에게는 KeplrError라는 타입으로 나올수도 있다는 걸 typing이 보여주는 효과가 있긴한데 뭐 살려도 되고 지워도 될듯요...

File: packages/router/src/error.ts (hunk 2 above)
Comment (junhoyeo) [reply packages/router/src/error.ts]: 추가 완료!

File: packages/extension/src/pages/main/account.tsx (hunk 4 above)
Comment (junhoyeo) [reply packages/extension/src/pages/main/account.tsx:67]: 좋아요 ㅎㅎ 위꺼랑 같이 추가했어요

**Pull request merged**